File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        curr_path = []
        count = 0
        def dfs(root: TreeNode):
            nonlocal count
            if root is None: return
            
            curr_path.append(root.val)
            maximum = max(curr_path)
            if root.val >= maximum: count += 1
            
            dfs(root.left)
            dfs(root.right)

            curr_path.pop()
        dfs(root)
        return count

File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    private:
        void dfs(TreeNode* root, std::vector<int>& inorder_vals) {
            if (!root) return;
            dfs(root->left, inorder_vals);
            inorder_vals.push_back(root->val);
            dfs(root->right, inorder_vals);
        }
    public:
        bool isValidBST(TreeNode* root) {
            std::vector<int> inorder_vals{};
            dfs(root, inorder_vals);
            
            for (int i = 1; i < inorder_vals.size(); ++i) {
                if (inorder_vals[i] <= inorder_vals[i - 1])
                    return false;
            }

            return true;
    }
};
